File: apps/agent-orchestrator/src/knowledge-base/retrieve.test.ts
import { describe, expect, it, vi } from "vitest";
import { BrokerProber, retrieve } from "./retrieve.js";
import { PermissionDeniedError, TransientProbeError, type ProbeResult, type Prober } from "./probe.js";
import type { CorpusQueryFilter, CorpusSearchResult, CorpusStore } from "./types.js";

function hit(sourceId: string, contentHash: string, score: number): CorpusSearchResult {
  return {
    score,
    chunk: {
      connectionId: "c",
      sourceId,
      contentHash,
      sourceUrl: "https://mirror.invalid/x",
      title: "MIRROR TITLE",
      version: "v1",
      text: "…",
    },
  };
}

class FakeStore implements CorpusStore {
  constructor(
    private readonly hits: CorpusSearchResult[] = [],
    private readonly failure?: Error,
  ) {}
  async query(_t: string, _f: CorpusQueryFilter, k: number) {
    if (this.failure) throw this.failure;
    return [...this.hits].sort((a, b) => b.score - a.score).slice(0, k);
  }
}

class FakeProber implements Prober {
  probes = 0;
  constructor(
    private readonly results: Record<string, ProbeResult> = {},
    private readonly errors: Record<string, Error> = {},
  ) {}
  granularity() {
    return "resource" as const;
  }
  async probe(request: { connectionId: string; sourceId?: string }) {
    this.probes += 1;
    const key = `${request.connectionId}/${request.sourceId ?? ""}`;
    const failure = this.errors[key];
    if (failure) throw failure;
    return this.results[key] ?? { allowed: false, title: "", url: "" };
  }
}

describe("retrieve", () => {
  it("returns only what the source confirmed", async () => {
    const store = new FakeStore([hit("readable", "h1", 0.9), hit("restricted", "h2", 0.8)]);
    const prober = new FakeProber(
      { "c/readable": { allowed: true, title: "Readable", url: "u1", version: "v1" } },
      { "c/restricted": new PermissionDeniedError("403") },
    );

    const outcome = await retrieve([store], prober, "q", ["reader"], 5);

    expect(outcome.chunks).toHaveLength(1);
    // The title came from the probe, not the mirror's "MIRROR TITLE".
    expect(outcome.chunks[0].title).toBe("Readable");
    expect(outcome.denied).toBe(1);
  });

  it("over-fetches so probe drops do not starve the answer", async () => {
    const many = Array.from({ length: 12 }, (_, i) => hit(`s${i}`, `h${i}`, (12 - i) / 12));
    const results: Record<string, ProbeResult> = {};
    for (let i = 0; i < 12; i += 1) {
      results[`c/s${i}`] = { allowed: true, title: "t", url: "u", version: "v1" };
    }
    const prober = new FakeProber(results);

    const outcome = await retrieve([new FakeStore(many)], prober, "q", ["reader"], 4, 3);

    expect(outcome.chunks).toHaveLength(4);
    expect(prober.probes).toBe(12);
  });

  it("falls back to the default multiplier for a nonsense one", async () => {
    const prober = new FakeProber({ "c/s": { allowed: true, title: "t", url: "u", version: "v1" } });
    const outcome = await retrieve([new FakeStore([hit("s", "h", 0.5)])], prober, "q", ["reader"], 2, 0);
    expect(outcome.chunks).toHaveLength(1);
  });

  it("carries both kinds of missing evidence", async () => {
    const healthy = new FakeStore([hit("busy", "h", 0.5)]);
    const broken = new FakeStore([], new Error("qdrant down"));
    const prober = new FakeProber({}, { "c/busy": new TransientProbeError("429") });

    const outcome = await retrieve([healthy, broken], prober, "q", ["reader"], 5);

    // A corpus that could not be searched and a source that could not be
    // checked are different failures, and an answer should be able to say both.
    expect(outcome.skippedCorpora).toBe(1);
    expect(outcome.undetermined).toEqual(["c/busy"]);
    expect(outcome.chunks).toEqual([]);
  });
});

describe("BrokerProber", () => {
  it("authenticates itself and forwards the user's own credential", async () => {
    const http = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      json: async () => ({ allowed: true, title: "Auth design", url: "https://wiki/1", version: "v7" }),
    } as unknown as Response);

    const prober = new BrokerProber({
      baseUrl: "http://broker",
      token: "orch-token",
      delegatedToken: "user-token",
      fetchImpl: http as unknown as typeof fetch,
    });

    const result = await prober.probe({ connectionId: "snc-confluence", sourceId: "page-1" });

    expect(result.title).toBe("Auth design");
    const [url, init] = http.mock.calls[0]!;
    expect(url).toBe("http://broker/connections/snc-confluence/probe");
    expect((init as RequestInit).headers).toMatchObject({
      authorization: "Bearer orch-token",
      "x-delegated-token": "user-token",
    });
  });

  it.each([
    [403, "denied"],
    [404, "denied"],
    [429, "transient"],
    [500, "transient"],
    [503, "transient"],
  ])("classifies %i as %s", async (status, kind) => {
    const http = vi.fn().mockResolvedValue({ ok: false, status } as unknown as Response);
    const prober = new BrokerProber({
      baseUrl: "http://broker",
      token: "t",
      delegatedToken: "u",
      fetchImpl: http as unknown as typeof fetch,
    });

    const expected = kind === "denied" ? PermissionDeniedError : TransientProbeError;
    await expect(prober.probe({ connectionId: "c", sourceId: "s" })).rejects.toBeInstanceOf(expected);
  });

  it("refuses to probe without a delegated credential", async () => {
    const prober = new BrokerProber({ baseUrl: "http://broker", token: "t" });
    await expect(prober.probe({ connectionId: "c", sourceId: "s" })).rejects.toBeInstanceOf(
      PermissionDeniedError,
    );
  });

  it("treats an unreachable broker as transient, not a denial", async () => {
    const http = vi.fn().mockRejectedValue(new Error("ECONNREFUSED"));
    const prober = new BrokerProber({
      baseUrl: "http://broker",
      token: "t",
      delegatedToken: "u",
      fetchImpl: http as unknown as typeof fetch,
    });

    await expect(prober.probe({ connectionId: "c", sourceId: "s" })).rejects.toBeInstanceOf(
      TransientProbeError,
    );
  });

  it("defaults an unknown provider to per-resource probing", () => {
    const prober = new BrokerProber({
      baseUrl: "http://broker",
      token: "t",
      granularities: new Map([["slack-eng", "connection" as const]]),
    });

    expect(prober.granularity("slack-eng")).toBe("connection");
    // Assuming per-connection would let one allowed resource vouch for every
    // other candidate from that source.
    expect(prober.granularity("something-new")).toBe("resource");
  });
});

File: engines/temporal/internal/corpus/retrieve_test.go
package corpus_test

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/controller-agent/temporal-engine/internal/corpus"
	"github.com/controller-agent/temporal-engine/internal/vectorstore"
)

func TestRetrieveReturnsOnlyWhatTheSourceConfirmed(t *testing.T) {
	store := &fakeStore{hits: []vectorstore.Hit{
		chunkHit("c", "readable", "h1", 0.9),
		chunkHit("c", "restricted", "h2", 0.8),
	}}
	prober := &fakeProber{
		results: map[string]corpus.ProbeResult{
			"c/readable": {Allowed: true, Title: "Readable", URL: "u1", Version: "v1"},
		},
		errs: map[string]error{
			"c/restricted": &corpus.PermissionDenied{Err: errors.New("403")},
		},
	}

	outcome, err := corpus.Retrieve(context.Background(),
		[]vectorstore.Store{store}, prober, "q", []string{"reader"}, nil, 5, 3)

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 1)
	require.Equal(t, "Readable", outcome.Chunks[0].Title)
	require.Equal(t, 1, outcome.Denied)
}

func TestRetrieveOverFetchesSoProbeDropsDoNotStarveTheAnswer(t *testing.T) {
	// Twelve candidates exist; the answer wants 4. Asking the store for exactly
	// 4 would leave nothing in reserve when the source refuses some of them.
	hits := make([]vectorstore.Hit, 0, 12)
	for i := 0; i < 12; i++ {
		hits = append(hits, chunkHit("c", string(rune('a'+i)), string(rune('a'+i)), float32(12-i)/12))
	}
	store := &fakeStore{hits: hits}

	prober := &fakeProber{results: map[string]corpus.ProbeResult{}}
	for i := 0; i < 12; i++ {
		prober.results["c/"+string(rune('a'+i))] = corpus.ProbeResult{
			Allowed: true, Title: "t", URL: "u", Version: "v1",
		}
	}

	outcome, err := corpus.Retrieve(context.Background(),
		[]vectorstore.Store{store}, prober, "q", []string{"reader"}, nil, 4, 3)

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 4, "capped at the requested limit")
	require.Equal(t, 12, prober.probeCount(), "but 4*3 candidates were probed")
}

func TestRetrieveDefaultsAnAbsurdMultiplier(t *testing.T) {
	store := &fakeStore{hits: []vectorstore.Hit{chunkHit("c", "s", "h", 0.5)}}
	prober := &fakeProber{results: map[string]corpus.ProbeResult{
		"c/s": {Allowed: true, Title: "t", URL: "u", Version: "v1"},
	}}

	outcome, err := corpus.Retrieve(context.Background(),
		[]vectorstore.Store{store}, prober, "q", []string{"reader"}, nil, 2, 0)

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 1)
}

func TestRetrieveCarriesBothKindsOfMissingEvidence(t *testing.T) {
	healthy := &fakeStore{hits: []vectorstore.Hit{chunkHit("c", "busy", "h", 0.5)}}
	broken := &fakeStore{err: errors.New("qdrant down")}
	prober := &fakeProber{errs: map[string]error{
		"c/busy": &corpus.Transient{Err: errors.New("429")},
	}}

	outcome, err := corpus.Retrieve(context.Background(),
		[]vectorstore.Store{healthy, broken}, prober, "q", []string{"reader"}, nil, 5, 3)

	require.NoError(t, err)
	// A corpus that could not be searched and a source that could not be
	// checked are different failures, and an answer should be able to say both.
	require.Equal(t, 1, outcome.SkippedCorpora)
	require.Equal(t, []string{"c/busy"}, outcome.Undetermined)
	require.Empty(t, outcome.Chunks)
}

func TestBrokerProberSendsTheDelegatedTokenAndReadsTheProbe(t *testing.T) {
	var gotAuth, gotDelegated, gotPath string
	var gotBody map[string]string

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("authorization")
		gotDelegated = r.Header.Get("x-delegated-token")
		gotPath = r.URL.Path
		_ = json.NewDecoder(r.Body).Decode(&gotBody)
		_ = json.NewEncoder(w).Encode(corpus.ProbeResult{
			Allowed: true, Title: "Auth design", URL: "https://wiki/1", Version: "v7",
		})
	}))
	defer server.Close()

	prober := &corpus.BrokerProber{
		BaseURL: server.URL, Token: "orch-token", DelegatedToken: "user-token",
	}

	result, err := prober.Probe(context.Background(),
		corpus.ProbeRequest{CorpusID: "globex-confluence", SourceID: "page-1"})

	require.NoError(t, err)
	require.Equal(t, "Auth design", result.Title)
	require.Equal(t, "Bearer orch-token", gotAuth, "the orchestrator authenticates itself")
	require.Equal(t, "user-token", gotDelegated, "and forwards the user's own credential")
	require.Equal(t, "/connections/globex-confluence/probe", gotPath)
	require.Equal(t, "page-1", gotBody["sourceId"])
}

func TestBrokerProberClassifiesBrokerStatuses(t *testing.T) {
	for _, tc := range []struct {
		status int
		denied bool
	}{
		{http.StatusForbidden, true},
		{http.StatusNotFound, true},
		{http.StatusTooManyRequests, false},
		{http.StatusServiceUnavailable, false},
		{http.StatusInternalServerError, false},
	} {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
			w.WriteHeader(tc.status)
		}))

		prober := &corpus.BrokerProber{BaseURL: server.URL, Token: "t", DelegatedToken: "u"}
		_, err := prober.Probe(context.Background(), corpus.ProbeRequest{CorpusID: "c", SourceID: "s"})

		var denied *corpus.PermissionDenied
		var transient *corpus.Transient
		if tc.denied {
			require.ErrorAs(t, err, &denied, "status %d must drop the candidate", tc.status)
		} else {
			// Treating a busy broker as a denial would quietly shrink the answer.
			require.ErrorAs(t, err, &transient, "status %d must NOT be read as a denial", tc.status)
		}
		server.Close()
	}
}

func TestBrokerProberRefusesToProbeWithoutADelegatedCredential(t *testing.T) {
	prober := &corpus.BrokerProber{BaseURL: "http://unused", Token: "t"}

	_, err := prober.Probe(context.Background(), corpus.ProbeRequest{CorpusID: "c", SourceID: "s"})

	// Nothing to answer the question with; the broker would refuse it anyway.
	var denied *corpus.PermissionDenied
	require.ErrorAs(t, err, &denied)
}

func TestBrokerProberTreatsAnUnreachableBrokerAsTransient(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(http.ResponseWriter, *http.Request) {}))
	url := server.URL
	server.Close() // nothing is listening now

	prober := &corpus.BrokerProber{BaseURL: url, Token: "t", DelegatedToken: "u"}
	_, err := prober.Probe(context.Background(), corpus.ProbeRequest{CorpusID: "c", SourceID: "s"})

	var transient *corpus.Transient
	require.ErrorAs(t, err, &transient)
}

func TestBrokerProberDefaultsToPerResourceGranularity(t *testing.T) {
	prober := &corpus.BrokerProber{
		Granularities: map[string]corpus.Granularity{"slack-eng": corpus.GranularityConnection},
	}

	require.Equal(t, corpus.GranularityConnection, prober.Granularity("slack-eng"))
	// Assuming per-connection for an unknown provider would let one allowed
	// resource vouch for every other candidate from that source.
	require.Equal(t, corpus.GranularityResource, prober.Granularity("something-new"))
}
